Add legacy protocol support

syssi opened the pull request:
@inorhan: Could you download this feature branch and flash the `esp32-example.yaml` to your esp32:

```
wget https://raw.githubusercontent.com/syssi/esphome-jk-bms/add-protocol-v4-support/esp32-example.yaml
esphome run esp32-example.yaml
```

I've added a single query to this feature branch. Your BMS should respond if it supports the protocol v4. You should see something like this:

```
[20:36:28][VV][scheduler:152]: Running interval 'update' with interval=5000 last_execution=4294966687 (now=4391)
[20:36:28][VV][uart.arduino_esp32:121]:     Wrote 0b11011101 (0xDD)
[20:36:28][VV][uart.arduino_esp32:121]:     Wrote 0b10100101 (0xA5)
[20:36:28][VV][uart.arduino_esp32:121]:     Wrote 0b00000011 (0x03)
[20:36:28][VV][uart.arduino_esp32:121]:     Wrote 0b00000000 (0x00)
[20:36:28][VV][uart.arduino_esp32:121]:     Wrote 0b11111111 (0xFF)
[20:36:28][VV][uart.arduino_esp32:121]:     Wrote 0b11111101 (0xFD)
[20:36:28][VV][uart.arduino_esp32:121]:     Wrote 0b01110111 (0x77)
[20:36:28][VV][uart.arduino_esp32:142]:     Flushing...
# The BMS responds should follow here...
```

Please provide some responses of your BMS.

Closes: #13

inorhan commented:
I receive the same messages being connected/disconnected to the BMS.

disconnected:

[17:54:57][VV][scheduler:152]: Running interval 'update' with interval=5000 last_execution=78911 (now=83911)
[17:54:57][VV][uart.arduino_esp32:121]:     Wrote 0b11011101 (0xDD)
[17:54:57][VV][uart.arduino_esp32:121]:     Wrote 0b10100101 (0xA5)
[17:54:57][VV][uart.arduino_esp32:121]:     Wrote 0b00000011 (0x03)
[17:54:57][VV][uart.arduino_esp32:121]:     Wrote 0b00000000 (0x00)
[17:54:57][VV][uart.arduino_esp32:121]:     Wrote 0b11111111 (0xFF)
[17:54:57][VV][uart.arduino_esp32:121]:     Wrote 0b11111101 (0xFD)
[17:54:57][VV][uart.arduino_esp32:121]:     Wrote 0b01110111 (0x77)
[17:54:57][VV][uart.arduino_esp32:142]:     Flushing...
[17:55:02][VV][scheduler:152]: Running interval 'update' with interval=5000 last_execution=83911 (now=88915)
[17:55:02][VV][uart.arduino_esp32:121]:     Wrote 0b11011101 (0xDD)
[17:55:02][VV][uart.arduino_esp32:121]:     Wrote 0b10100101 (0xA5)
[17:55:02][VV][uart.arduino_esp32:121]:     Wrote 0b00000011 (0x03)
[17:55:02][VV][uart.arduino_esp32:121]:     Wrote 0b00000000 (0x00)
[17:55:02][VV][uart.arduino_esp32:121]:     Wrote 0b11111111 (0xFF)
[17:55:02][VV][uart.arduino_esp32:121]:     Wrote 0b11111101 (0xFD)
[17:55:02][VV][uart.arduino_esp32:121]:     Wrote 0b01110111 (0x77)
[17:55:02][VV][uart.arduino_esp32:142]:     Flushing...


connected:

[17:58:32][VV][scheduler:152]: Running interval 'update' with interval=5000 last_execution=94285 (now=99290)
[17:58:32][VV][uart.arduino_esp32:121]:     Wrote 0b11011101 (0xDD)
[17:58:32][VV][uart.arduino_esp32:121]:     Wrote 0b10100101 (0xA5)
[17:58:32][VV][uart.arduino_esp32:121]:     Wrote 0b00000011 (0x03)
[17:58:32][VV][uart.arduino_esp32:121]:     Wrote 0b00000000 (0x00)
[17:58:32][VV][uart.arduino_esp32:121]:     Wrote 0b11111111 (0xFF)
[17:58:32][VV][uart.arduino_esp32:121]:     Wrote 0b11111101 (0xFD)
[17:58:32][VV][uart.arduino_esp32:121]:     Wrote 0b01110111 (0x77)
[17:58:32][VV][uart.arduino_esp32:142]:     Flushing...
[17:58:37][VV][scheduler:152]: Running interval 'update' with interval=5000 last_execution=99285 (now=104286)
[17:58:37][VV][uart.arduino_esp32:121]:     Wrote 0b11011101 (0xDD)
[17:58:37][VV][uart.arduino_esp32:121]:     Wrote 0b10100101 (0xA5)
[17:58:37][VV][uart.arduino_esp32:121]:     Wrote 0b00000011 (0x03)
[17:58:37][VV][uart.arduino_esp32:121]:     Wrote 0b00000000 (0x00)
[17:58:37][VV][uart.arduino_esp32:121]:     Wrote 0b11111111 (0xFF)
[17:58:37][VV][uart.arduino_esp32:121]:     Wrote 0b11111101 (0xFD)
[17:58:37][VV][uart.arduino_esp32:121]:     Wrote 0b01110111 (0x77)
[17:58:37][VV][uart.arduino_esp32:142]:     Flushing...
[17:58:37][V][component:186]: Component esphome.coroutine took a long time for an operation (0.05 s).
[17:58:37][V][component:187]: Components should block for at most 20-30ms.


I'm using the original RS232-485 JKBMS converter + a XY-017 RS485-TTL converter to the esp:

  platform: ESP32
  board: nodemcu-32s

  baud_rate: 9600
  rx_buffer_size: 384
  tx_pin: GPIO14
  rx_pin: GPIO4

syssi commented:
Does the TX led of the rs485 module blink periodically? Please make sure the TX led is blinking and not the RX led. Did you try to swap the RS485 (A/B) wires once?

inorhan commented:
Tx blinks periodically. Rx stays off.
I will try swapping them either way.

El lun., 1 nov. 2021 19:51, Sebastian Muszynski ***@***.***>
escribió:

> Does the TX led of the rs485 module blink periodically? Please make sure
> the TX led is blinking and not the RX led. Did you try to swap the RS485
> (A/B) wires once?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/syssi/esphome-jk-bms/pull/14#issuecomment-956497286>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACYQ3FRRP3QUWEXGI2XBIK3UJ3OTDANCNFSM5HCUHQOQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

inorhan commented:
Tested swapping A/B wires for both the new and normal esphome-jk-bms. Still
the same.

El mar, 2 nov 2021 a las 0:57, inorhan inorhan ***@***.***>)
escribió:

> Tx blinks periodically. Rx stays off.
> I will try swapping them either way.
>
> El lun., 1 nov. 2021 19:51, Sebastian Muszynski ***@***.***>
> escribió:
>
>> Does the TX led of the rs485 module blink periodically? Please make sure
>> the TX led is blinking and not the RX led. Did you try to swap the RS485
>> (A/B) wires once?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/syssi/esphome-jk-bms/pull/14#issuecomment-956497286>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACYQ3FRRP3QUWEXGI2XBIK3UJ3OTDANCNFSM5HCUHQOQ>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>>
>

syssi commented:
Could you try to swap the GPIOs at the yaml once. Does the RX led blink this time? I want to make sure the GPIOs are connected properly. You have to swap the GPIOs later again because the TX led must blink / the ESPs sends periodically.

The last step would be dropping both rs485 converters. You can connect the ESP directly to the BMS because of the TTL level (3.3V). This time it is super important to identify the RX and TX pin of the BMS. Don't use `VBAT` by mistake. You could measure the voltage between the outer pins to identify VBAT and GND. If the measured voltage is positive you now the exact positions. If the voltage is negative you have to swap the voltmeter probes.

inorhan commented:
Can i just change the code in esphome ? Pin4 <->14


El mar., 2 nov. 2021 8:28, Sebastian Muszynski ***@***.***>
escribió:

> Could you try to swap the GPIOs at the yaml once. Does the RX led blink
> this time? I want to make sure the GPIOs are connected properly. You have
> to swap the GPIOs later again because the TX led must blink / the ESPs
> sends periodically.
>
> The last step would be dropping both rs485 converters. You can connect the
> ESP directly to the BMS because of the TTL level (3.3V). This time it is
> super important to identify the RX and TX pin of the BMS. Don't use VBAT
> by mistake. You could measure the voltage between the outer pins to
> identify VBAT and GND. If the measured voltage is positive you now the
> exact positions. If the voltage is negative you have to swap the voltmeter
> probes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/syssi/esphome-jk-bms/pull/14#issuecomment-957163430>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACYQ3FSNWP4ZSYUCNGV6HXTUJ6HITANCNFSM5HCUHQOQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

syssi commented:
> Can i just change the code in esphome? Pin4 <->14

Yes. This is the idea! So we can be sure the GPIOs are properly connected (without re-wiring).

inorhan commented:
1) Original code:

A & B no swap / A & B swap= All the esphome sensors remains 0
TX LED in the 485 converter flickers as usual, RX off
tx_pin: GPIO14
rx_pin: GPIO4

2) Original code:

  baud_rate: 115200
  rx_buffer_size: 384
  tx_pin: GPIO4  #original 14
  rx_pin: GPIO14   #original 4

A & B swapp or A & B normal:
TX LED in the 485 converter flickers as usual, RX off
Even with the BMS cable disconnected I get the same garbage data .. must be
reading itself.

esphome_jkbms temperature sensor 1 -22289 °C
esphome_jkbms temperature sensor 2 -22289 °C
esphome_jkbms total voltage 403.07 V
esphome_jkbms current 0.00 A
esphome_jkbms capacity remaining 253 %
esphome_jkbms charging cycles 64831
esphome_jkbms total charging cycle capacity 1945976576
esphome_jkbms total voltage overvoltage protection 163.04 V
esphome_jkbms total voltage undervoltage protection 648.31 V
esphome_jkbms cell voltage overvoltage protection 22.189 V
esphome_jkbms cell voltage overvoltage recovery 51.315 V
esphome_jkbms cell voltage overvoltage delay 16180 s
esphome_jkbms cell voltage undervoltage protection 47.787 V
esphome_jkbms cell voltage undervoltage recovery 0.000 V
esphome_jkbms cell voltage undervoltage delay 43130 s
esphome_jkbms cell pressure difference protection 16.248 V
esphome_jkbms balance starting voltage 0.000 V
esphome_jkbms balance opening pressure difference 30.205 V
esphome_jkbms total runtime 24237100 h
esphome_jkbms actual battery capacity 855668800 Ah
esphome_jkbms charging ON
esphome_jkbms discharging
esphome_jkbms balancing
esphome_jkbms operation mode Balancer enabled


3) test code:

uart:
  id: uart0
  baud_rate: 9600
  rx_buffer_size: 384
  tx_pin: GPIO4  #original 14
  rx_pin: GPIO14   #original 4

TX LED in the 485 converter flickers as usual, RX always off
Even with the BMS cable disconnected I get the same garbage data .. must be
reading itself.

[13:01:00][VV][uart.arduino_esp32:121]:     Wrote 0b11011101 (0xDD)
[13:01:00][VV][uart.arduino_esp32:121]:     Wrote 0b10100101 (0xA5)
[13:01:00][VV][uart.arduino_esp32:121]:     Wrote 0b00000011 (0x03)
[13:01:00][VV][uart.arduino_esp32:121]:     Wrote 0b00000000 (0x00)
[13:01:00][VV][uart.arduino_esp32:121]:     Wrote 0b11111111 (0xFF)
[13:01:00][VV][uart.arduino_esp32:121]:     Wrote 0b11111101 (0xFD)
[13:01:00][VV][uart.arduino_esp32:121]:     Wrote 0b01110111 (0x77)
[13:01:00][VV][uart.arduino_esp32:142]:     Flushing...
[13:01:00][V][component:186]: Component esphome.coroutine took a long time
for an operation (0.06 s).
[13:01:00][V][component:187]: Components should block for at most 20-30ms.
[13:01:00][VV][uart.arduino_esp32:135]:     Read 0b11011101 (0xDD)
[13:01:00][VV][uart.arduino_esp32:135]:     Read 0b10100101 (0xA5)
[13:01:00][VV][uart.arduino_esp32:135]:     Read 0b00000011 (0x03)
[13:01:00][VV][uart.arduino_esp32:135]:     Read 0b00000000 (0x00)
[13:01:00][VV][uart.arduino_esp32:135]:     Read 0b11111111 (0xFF)
[13:01:00][VV][uart.arduino_esp32:135]:     Read 0b11111101 (0xFD)
[13:01:00][VV][uart.arduino_esp32:135]:     Read 0b01110111 (0x77)
[13:01:00][V][component:186]: Component jk_modbus took a long time for an
operation (0.05 s).
[13:01:00][V][component:187]: Components should block for at most 20-30ms.
[13:01:05][VV][scheduler:152]: Running interval 'update' with interval=5000
last_execution=139366 (now=144366)
[13:01:05][VV][uart.arduino_esp32:121]:     Wrote 0b11011101 (0xDD)
[13:01:05][VV][uart.arduino_esp32:121]:     Wrote 0b10100101 (0xA5)
[13:01:05][VV][uart.arduino_esp32:121]:     Wrote 0b00000011 (0x03)
[13:01:05][VV][uart.arduino_esp32:121]:     Wrote 0b00000000 (0x00)
[13:01:05][VV][uart.arduino_esp32:121]:     Wrote 0b11111111 (0xFF)
[13:01:05][VV][uart.arduino_esp32:121]:     Wrote 0b11111101 (0xFD)
[13:01:05][VV][uart.arduino_esp32:121]:     Wrote 0b01110111 (0x77)
[13:01:05][VV][uart.arduino_esp32:142]:     Flushing...
[13:01:05][VV][uart.arduino_esp32:135]:     Read 0b11011101 (0xDD)
[13:01:05][VV][uart.arduino_esp32:135]:     Read 0b10100101 (0xA5)
[13:01:05][VV][uart.arduino_esp32:135]:     Read 0b00000011 (0x03)
[13:01:05][VV][uart.arduino_esp32:135]:     Read 0b00000000 (0x00)
[13:01:05][VV][uart.arduino_esp32:135]:     Read 0b11111111 (0xFF)
[13:01:05][VV][uart.arduino_esp32:135]:     Read 0b11111101 (0xFD)
[13:01:05][VV][uart.arduino_esp32:135]:     Read 0b01110111 (0x77)


I don't get why the TX LED always flickers (every 2 seconds aprox) in the
rs485 converter, even changing 4<->14.
RX stays off all the time.

El mar, 2 nov 2021 a las 12:49, Sebastian Muszynski (<
***@***.***>) escribió:

> Can i just change the code in esphome? Pin4 <->14
>
> Yes. This is the idea! So we can be sure the GPIOs are properly connected
> (without re-wiring).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/syssi/esphome-jk-bms/pull/14#issuecomment-957365463>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACYQ3FW7VMVE6CKRY7C6IS3UJ7F23ANCNFSM5HCUHQOQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

syssi commented:
This is super strange. I talked to some aliexpress seller of the BMS. He told me something like: Every BMS with firmware version `<=6.x` uses the old protocol (request: `0xDD 0xA5 0x03 0x00 0xFF 0xFD 0x77`). A BMS with a firmware version >6 (f.e. 7.x and 8.x) should use the new protocol. And it looks like the baudrate has changed over time, so you should test 9600 bauds and 115200 bauds. 

IMO your device should work with the original code of this component. Could you try to connect the ESP32 directly to your BMS?

inorhan commented:
Connected directly (no 485 converters), with original code, swapping/ not
swapping 4 / 14 pin, 9600/115200 bps.

I always get crazy data in the sensors.
Connecting or disconnecting the com cable to the BMS, doesn't change
anything, still receiving the same sensor data.
Also, I noticed that the voltage between GND and VBAT in the JST connector
=0 vdc.  Is this normal for the 485/GPS port ?

[18:18:36][I][jk_bms:061]: Status frame received [18:18:36][D][sensor:113]:
'esphome_jkbms temperature sensor 1': Sending state -22289.00000 °C with 0
decimals of accuracy [18:18:36][D][sensor:113]: 'esphome_jkbms temperature
sensor 2': Sending state -22289.00000 °C with 0 decimals of accuracy
[18:18:36][D][sensor:113]: 'esphome_jkbms total voltage': Sending state
308.06000 V with 2 decimals of accuracy [18:18:36][D][sensor:113]:
'esphome_jkbms current': Sending state 0.00000 A with 2 decimals of accuracy
[18:18:36][D][sensor:113]: 'esphome_jkbms capacity remaining': Sending
state 253.00000 % with 0 decimals of accuracy [18:18:36][D][sensor:113]:
'esphome_jkbms charging cycles': Sending state 47787.00000 with 0 decimals
of accuracy [18:18:36][D][sensor:113]: 'esphome_jkbms total charging cycle
capacity': Sending state 0.00000 with 0 decimals of accuracy
[18:18:36][D][text_sensor:067]:
'esphome_jkbms errors': Sending state 'Low capacity;Charging
overvoltage;Discharging undervoltage;Battery over temperature;Charging
overcurrent;Discharging overcurrent;Cell pressure difference;Battery low
temperature' [18:18:36][D][text_sensor:067]: 'esphome_jkbms operation
mode': Sending state '' [18:18:36][D][sensor:113]: 'esphome_jkbms total
voltage overvoltage protection': Sending state 0.00000 V with 2 decimals of
accuracy [18:18:36][D][sensor:113]: 'esphome_jkbms total voltage
undervoltage protection': Sending state 0.00000 V with 2 decimals of
accuracy [18:18:36][D][sensor:113]: 'esphome_jkbms cell voltage overvoltage
protection': Sending state 0.00000 V with 3 decimals of accuracy
[18:18:36][D][sensor:113]:
'esphome_jkbms cell voltage overvoltage recovery': Sending state 0.00000 V
with 3 decimals of accuracy [18:18:36][D][sensor:113]: 'esphome_jkbms cell
voltage overvoltage delay': Sending state 0.00000 s with 0 decimals of
accuracy [18:18:36][D][sensor:113]: 'esphome_jkbms cell voltage
undervoltage protection': Sending state 0.00000 V with 3 decimals of
accuracy [18:18:36][D][sensor:113]: 'esphome_jkbms cell voltage
undervoltage recovery': Sending state 0.00000 V with 3 decimals of accuracy
[18:18:36][D][sensor:113]: 'esphome_jkbms cell voltage undervoltage delay':
Sending state 0.00000 s with 0 decimals of accuracy [18:18:36][D][sensor:113]:
'esphome_jkbms cell pressure difference protection': Sending state 0.00000
V with 3 decimals of accuracy [18:18:36][D][sensor:113]: 'esphome_jkbms
balance starting voltage': Sending state 0.00000 V with 3 decimals of
accuracy [18:18:36][D][sensor:113]: 'esphome_jkbms balance opening pressure
difference': Sending state 0.00000 V with 3 decimals of accuracy
[18:18:36][D][sensor:113]:
'esphome_jkbms total runtime': Sending state 5237056.00000 h with 0
decimals of accuracy [18:18:36][D][sensor:113]: 'esphome_jkbms actual
battery capacity': Sending state 2018946432.00000 Ah with 0 decimals of
accuracy [18:18:41][I][jk_bms:061]: Status frame received
[18:18:41][D][sensor:113]:
'esphome_jkbms temperature sensor 1': Sending state -22289.00000 °C with 0
decimals of accuracy [18:18:41][D][sensor:113]: 'esphome_jkbms temperature
sensor 2': Sending state -22289.00000 °C with 0 decimals of accuracy

El mar, 2 nov 2021 a las 16:23, Sebastian Muszynski (<
***@***.***>) escribió:

> This is super strange. I talked to some aliexpress seller of the BMS. He
> told me something like: Every BMS with firmware version <=6.x uses the
> old protocol (request: 0xDD 0xA5 0x03 0x00 0xFF 0xFD 0x77). A BMS with a
> firmware version >6 (f.e. 7.x and 8.x) should use the new protocol. And it
> looks like the baudrate has changed over time, so you should test 9600
> bauds and 115200 bauds.
>
> IMO your device should work with the original code of this component.
> Could you try to connect the ESP32 directly to your BMS?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/syssi/esphome-jk-bms/pull/14#issuecomment-957812947>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACYQ3FRZSG73BDX4ROTOCETUJ766VANCNFSM5HCUHQOQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

syssi commented:
How many cables did you connect to the BMS?

inorhan commented:
gnd, tx,rx  (tested also without connecting gnd to the ESP).
I'm powering the ESP from an isolated 5v PSU.
Never used Vbat (red cable from the JST),

El mar, 2 nov 2021 a las 19:05, Sebastian Muszynski (<
***@***.***>) escribió:

> How many cables did you connect to the BMS?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/syssi/esphome-jk-bms/pull/14#issuecomment-957998956>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACYQ3FXLQVYROL7YSUQY3RLUKAR5VANCNFSM5HCUHQOQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

syssi commented:
A side note: If your BMS doesn't provide VBAT the original RS485 adapter isn't powered properly and cannot work.

inorhan commented:
You are right, Vbat wasn't connected to the multimeter before...
Vbat provides battery voltage correctly.

El mar, 2 nov 2021 a las 19:23, Sebastian Muszynski (<
***@***.***>) escribió:

> A side note: If your BMS doesn't provide VBAT the original RS485 adapter
> isn't powered properly and cannot work.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/syssi/esphome-jk-bms/pull/14#issuecomment-958015391>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACYQ3FRYNRBOQDTF4ZCQM5TUKAUDLANCNFSM5HCUHQOQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

inorhan commented:
BTW, there is another github that actually works with the new versions of
JKBMS:
https://github.com/Louisvdw/dbus-serialbattery/blob/master/etc/dbus-serialbattery/jkbms.py
¿Is this the same protocol that you have implemented ?

El mar, 2 nov 2021 a las 19:47, inorhan inorhan ***@***.***>)
escribió:

> You are right, Vbat wasn't connected to the multimeter before...
> Vbat provides battery voltage correctly.
>
> El mar, 2 nov 2021 a las 19:23, Sebastian Muszynski (<
> ***@***.***>) escribió:
>
>> A side note: If your BMS doesn't provide VBAT the original RS485 adapter
>> isn't powered properly and cannot work.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/syssi/esphome-jk-bms/pull/14#issuecomment-958015391>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACYQ3FRYNRBOQDTF4ZCQM5TUKAUDLANCNFSM5HCUHQOQ>
>> .
>> Triage notifications on the go with GitHub Mobile for iOS
>> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
>> or Android
>> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>>
>>
>

syssi commented:
Yes. It's the same protocol. The is the command to query the current status:

https://github.com/Louisvdw/dbus-serialbattery/blob/master/etc/dbus-serialbattery/jkbms.py#L19

This component does the same here:

https://github.com/syssi/esphome-jk-bms/blob/main/components/jk_modbus/jk_modbus.cpp#L126-L147

syssi commented:
Could you try to use the python implementation by connecting the BMS to a TTL-to-USB adapter?

syssi committed:
Add jk_bms_old_ble component for the legacy (pre-6.0) JBD-style BLE protocol

Firmware <6.0 JK-BMS units respond to a DD/A5/77-framed basic-info request
over the JBD/Xiaoxiang-compatible BLE-UART bridge (service 0xFFE0) instead
of the current jk_bms_ble frame format. Adds a read-only component decoding
total voltage, current, capacity, cycles, SoC, cell count, temperatures and
charging/discharging MOSFET state, based on the real capture from #13.

Closes #13, supersedes the abandoned #14.
